### ISSUE TITLE
0.2.9 CONTAINS BREAKING CHANGES!!!

### ISSUE THREAD

**Issue by jd-carroll** [typeorm/typeorm#3159]
The recent 0.2.9 release has a single changelog entry of: 
```
* UpdateEvent now returns with contains updatedColumns and updatedRelations
```

However, there are a great many set of changes between the two versions.  So I am **very confident** that 0.2.9 contains changes that were not supposed to be released.

**Breaking Change:**
In 0.2.8, the `DriverUtils` will parse any URL provided and merge the contents from the URL with the config strategy using the code:
```javascript
  return Object.assign(options, urlDriverOptions);
```

In 0.2.9, the `DriverUtils` changes this approach to merge the parsed options to:
```javascript
  return Object.assign(urlDriverOptions, options);
```

The difference is that if properties like `username` and `password` are specified in the configuration object directly **and not in the url**, they will now be `undefined`.

Whether or not this was intentional, it at least needs to be called out more explicitly in the changelog

**Comment by havenchyk** [comment]
@jd-carroll thank you! Would you mind to provide a PR for that issue you found?

**Comment by jd-carroll** [comment]
@havenchyk Before submitting a PR, which I am happy to do, I would like to first confirm what is and is not supposed to be in the 0.2.9 release.  Since there isn't a `0.2.8` or `0.2.9` tag, I would recommend npm install'ing to two empty directories and performing a diff on the file system.

**Comment by havenchyk** [comment]
@jussikinnula could you please clarify this change?

**Comment by jussikinnula** [comment]
> @jussikinnula could you please clarify this change?

Sorry of the confusion. I did had to double check, and the fix works on my test application with both ConnectionOptions `url` set and with regular parameters (e.g. `username`, `password`, etc). If providing both, then the parameters are overwritten those in the URL, e.g.:

```
const options: ConnectionOptions = {
  url: 'postgres://user1:pass2@localhost:5432/db1',
  type: 'mysql',
  host: 'localhost',
  port: 3306,
  username: 'user1',
  password: 'user2',
  database: 'db2',    
}
```

This will get end result of `type` being `mysql`, but rest of the parameters are extracted from url. I guess this is undesirable?

Anyway, without the change `return Object.assign(urlDriverOptions, options);` none of the URL parameters will get through.

A working test application is here: https://github.com/jussikinnula/react-typeorm-todo

Just run `USE_TYPEORM_URL=true TYPEORM_URL=xxxx npm run dev` or `npm run dev` (and edit `src/server/typeOrmConfig.ts` to see how it works).

If we collectively agree, of course we can revert the original change - but I would see that the issue of `options.url` not working at all would then come back. If you want I can try to fix it some other way around. It would require a bit more heavier modifications of the logic in parsing configuration, unfortunately (I actually did my first version like that, but didn't want to propose that big changes).

On real life situation I see that modern apps are ran on cloud systems, which use 12-factor App principle. So everything is in environment variables. Of course database URL's could be parsed manually and then use normal parameters, but then it would make sense to remove `options.url` completely (if it's not desired to be used).

Any thoughts?

The current implementation with the "breaking change" doesn't break anything if you don't use `options.url`. Of course it should be documented that `options.url` overrides stuff. I agree that the current implementation is unoptimal, as it takes parameters from both `options.url` and from other parameters as well.

**Comment by jussikinnula** [comment]
An easy fix that I can think would be to strip all keys with undefined or null keys from options, then assign options from url. I can make a patch if this is viable

**Comment by havenchyk** [comment]
@jussikinnula I didn't dig into the code yet, but just to clarify, is the problem in overriding of some keys from config with values from url? like password and username? 

If so, I would suggest to add a note in docs and probably show a warning in console that both options are used and url will overwrite other keys

**Comment by havenchyk** [comment]
@jussikinnula thanks for the explanation. Your solution seems to be fine and working, but it's magical to me. I mean It will parse both URL and options and mix them, I don't know what was the initial idea (URL or options should have higher priority), but currently, I would add a warning if a user uses both URL and options with an explanation, how the values are used. Maybe with a link to docs.

What do you think @jussikinnula @pleerock ?

**Comment by jussikinnula** [comment]
> What do you think @jussikinnula @pleerock ?

It's perfectly possible for people to want to use URL and parameters manually. For example JDBC URL's don't contain username/password on the protocol format, which would make them have to input username and password manually.

Of course just a warning would not be bad idea, however, most of the cases the users would use either URL or with regular parameters.

Many libraries do use a specific method for connecting with URL to avoid confusion, and then another method to override URL parameters.

**Comment by pleerock** [comment]
fixed by #3173
